### PR TITLE
[RA] Enable LAN play.

### PR DESCRIPTION
**Pull request by OmniBlade**
Restores/Implements networking dialogs and networking code. Implements a pure sockets implementation for cross platform use.

### CODE REVIEW

**Comment by OmniBlade** [comment]
In theory I think this is ready for review, I can't see any code still missing. I've not got any way to test it set up locally at the moment though if anyone can test if its actually possible to join a created game and get into game with it.

File: redalert/wsproto.cpp [hunk 1]
```diff
@@ -57,6 +57,17 @@
 
 #include <stdio.h>
 
+WinsockInterfaceClass* PacketTransport = nullptr; // The object for interfacing with Winsock
+
+void Process_Network()
+{
+#if !defined _WIN32 || defined SDL2_BUILD
```
Comment (giulianobelinassi) [inline redalert/wsproto.cpp:64]: Why is it checking if SDL is enabled?

File: redalert/wsproto.cpp [hunk 2]
```diff
@@ -76,7 +87,12 @@
 WinsockInterfaceClass::WinsockInterfaceClass(void)
 {
     WinsockInitialised = false;
+#if defined _WIN32 && !defined SDL2_BUILD
```
Comment (giulianobelinassi) [inline redalert/wsproto.cpp:90]: Same thing here. Why is it checking if SDL is enabled?

File: redalert/wspudp.cpp [hunk 3]
```diff
@@ -209,19 +236,51 @@ bool UDPInterfaceClass::Open_Socket(SOCKET)
...
+
+    for (struct ifaddrs* ifa = if_addr; ifa != NULL; ifa = ifa->ifa_next) {
+        if (!ifa->ifa_addr) {
+            continue;
+        }
+        if (ifa->ifa_addr->sa_family == AF_INET) {
+            struct in_addr* tmp_addr = &((struct sockaddr_in*)ifa->ifa_addr)->sin_addr;
+            char buf[INET_ADDRSTRLEN];
+            inet_ntop(AF_INET, tmp_addr, buf, INET_ADDRSTRLEN);
+            char temp[128];
```
Comment (giulianobelinassi) [inline redalert/wspudp.cpp:257]: Can't an attacker overflow this buffer is the package is modified?

File: redalert/wsproto.cpp [hunk 4]
```diff
@@ -287,49 +321,37 @@ bool WinsockInterfaceClass::Init(void)
...
         OutputDebugString(out);
-        delete[] buffer;
-        return (false);
-    }
-
-    /*
-    ** Check the Winsock version number
-    */
-    if ((winsock_info->wVersion & 0x00ff) != (version & 0x00ff) || (winsock_info->wVersion >> 8) != (version >> 8)) {
-        OutputDebugString("TS: Winsock version is less than 1.1\n");
```
Comment (giulianobelinassi) [inline redalert/wsproto.cpp:322]: RA95: Winsock version...

**Comment by giulianobelinassi** [comment]
I tested it on Linux x86_64. Lobby works fine, ingame freezes when I move one unit.

Furthermore if two instances of vanillara is running on the same machine, once one join in the lobby, the second simply fails back to the main menu. I assume because `bind` failed, but I did not check this. It would be nice to have it working because it makes debugging easier.

**Comment by giulianobelinassi** [comment]
Stacktrace where it freezes:

```
#0  0x00007ffff7a758b3 in __GI___select (nfds=18, readfds=0x5555567d7720, writefds=0x5555567d77a0, exceptfds=0x0, timeout=0x7fffffffd820) at ../sysdeps/unix/sysv/linux/select.c:41
#1  0x00005555556d2bb2 in UDPInterfaceClass::Message_Handler (this=0x5555567d75d0) at /home/giulianob/bagunca/OmniBlade/Vanilla-Conquer/redalert/wspudp.cpp:508
#2  0x00005555556d1128 in Process_Network () at /home/giulianob/bagunca/OmniBlade/Vanilla-Conquer/redalert/wsproto.cpp:66
#3  0x00005555556da3d9 in WWKeyboardClass::Fill_Buffer_From_System (this=0x555555a6fe70) at /home/giulianob/bagunca/OmniBlade/Vanilla-Conquer/common/wwkeyboard.cpp:537
#4  0x00005555556d9e9a in WWKeyboardClass::Check (this=0x555555a6fe70) at /home/giulianob/bagunca/OmniBlade/Vanilla-Conquer/common/wwkeyboard.cpp:153
#5  0x0000555555661a28 in Wait_For_Players (first_time=0, net=0x555555a10160 <Ipx>, resend_delta=48, dialog_time=180, timeout=1800, 
    multi_packet_buf=0x555555a0fe50 <Session+3952> "\026>T\372<", my_sent=1, their_frame=0x555555a167e0 <Queue_AI_Multiplayer()::their_frame>, 
    their_sent=0x555555a16818 <Queue_AI_Multiplayer()::their_sent>, their_recv=0x555555a16828 <Queue_AI_Multiplayer()::their_recv>)
    at /home/giulianob/bagunca/OmniBlade/Vanilla-Conquer/redalert/queue.cpp:954
#6  0x00005555556616cf in Queue_AI_Multiplayer () at /home/giulianob/bagunca/OmniBlade/Vanilla-Conquer/redalert/queue.cpp:799
#7  0x0000555555661043 in Queue_AI () at /home/giulianob/bagunca/OmniBlade/Vanilla-Conquer/redalert/queue.cpp:377
#8  0x00005555555d0939 in Main_Loop () at /home/giulianob/bagunca/OmniBlade/Vanilla-Conquer/redalert/conquer.cpp:1851
#9  0x00005555555cd2d2 in Main_Game (argc=1, argv=0x7fffffffe058) at /home/giulianob/bagunca/OmniBlade/Vanilla-Conquer/redalert/conquer.cpp:309
#10 0x0000555555694f15 in main (argc=1, argv=0x7fffffffe058) at /home/giulianob/bagunca/OmniBlade/Vanilla-Conquer/redalert/startup.cpp:363

```

File: redalert/wspudp.cpp [hunk 5]
```diff
@@ -427,5 +488,116 @@ long UDPInterfaceClass::Message_Handler(HWND, UINT message, UINT, LONG lParam)
...
+
+    tv.tv_sec = 0;
+    tv.tv_usec = 0;
+
+    /*
+    ** Check if socket has data it would like to give us or can accept any data
+    */
+    fd_set* read_set = Get_Readset();
+    fd_set* write_set = Get_Writeset();
+    rc = select(Socket + 1, read_set, write_set, nullptr, &tv);
```
Comment (giulianobelinassi) [inline redalert/wspudp.cpp:508]: The read and write sets seems wrong here.

```
(gdb) p read_set[0]
$19 = {fds_bits = {0 <repeats 16 times>}}
(gdb) p read_set[1]
$20 = {fds_bits = {131072, 0 <repeats 15 times>}}
(gdb) p read_set[2]
$21 = {fds_bits = {94305939204608, 32, 3458764917452324374, 0, 393216, 0 <repeats 11 times>}}
(gdb) p read_set[3]
$22 = {fds_bits = {0 <repeats 16 times>}}
(gdb) p read_set[4]
$23 = {fds_bits = {0, 0, 0, 0, 93825011556064, 140737349151712, -711817106733197049, 33, 93825011565216, 140737349151712, 64, 48, 93825011565056, 0, 0, 0}}
(gdb) p write_set[4]
$24 = {fds_bits = {0, 769, 93825011556064, 140737349151712, -1882257339673417956, 33, 93825011565328, 140737349151712, 64, 48, 93825011565168, 0, 0, 0, 0, 657}}
(gdb) p write_set[0]
$25 = {fds_bits = {131072, 0 <repeats 15 times>}}
(gdb) p write_set[1]
$26 = {fds_bits = {94305939204608, 32, 3458764917452324374, 0, 393216, 0 <repeats 11 times>}}
(gdb) p write_set[2]
$27 = {fds_bits = {0 <repeats 16 times>}}
(gdb) p Socket
$28 = 17
(gdb) 
```

**Comment by giulianobelinassi** [comment]
Working on Linux 32-bits. Disconnects after some seconds on Linux 64-bits.

**Review by hifi — APPROVED** [review]
This PR is on the heavier side and has gone significant revisions but after @giulianobelinassi investigated and fixed the remaining 64-bit issues on Linux I'm happy to approve this after I did some quick tests :tada:

**Pull request merged**